config.yml

```diff
@@ -596,6 +596,8 @@ nodes:
         type: location
       - name: closing_loc
         type: location
+      - name: numbered_parameters
```

Comment:
With this configuration, consumers have to look at `numbered_parameters` every time they want to understand block/lambda nodes, and they have to implicitly know that if it's `> 0` then they should use the numbers and not check the regular parameters.

Another approach would be to create a `NumberedParametersNode`, which we put in place of parameters (which we already have the pointer for). We can do this because numbered parameters/other parameters are mutually exclusive. That way we wouldn't increase the size of every block/lambda node in the parser by 4 bytes and would instead only allocate for numbered parameters when necessary. Personally I think this is more clear to the consumer, because they don't have to know to check `numbered_parameters`. What do you think?

Reply:
Agreed a node would be nicer, as already suggested in https://github.com/ruby/prism/issues/1762#issuecomment-1816204931
Wouldn't it just work without any special casing if it was a BlockParametersNode too with an array of `RequiredParameterNode` (or `NumberedParameterNode`) with names _1, _2, etc?

Reply:
I initially thought about something like a `NumberedParametersNode`, but my hesitation was that it's not actually meaningful in terms of anything that's in the parameters. Usually the parameters node on something like a block or a lambda points concretely to the parameters actually being passed. Since numbered parameters are implied, they're never passed. A node like that would have, for instance, a null location, and somewhat arbitrary values (would we store, for instance, all of the numbered parameters that exist? the max one?). Is there prior art for this type of thing?

Reply:
@eregon you're welcome to add that to the desugar visitor, but that's a ton of memory that I don't want to allocate every time someone uses numbered parameters.

Reply:
My initial reaction to this PR is that it merges two concepts/types into one structure.  

As @kddnewton said it feels strange to check if numbered_parameters is checked to know whether you are dealing with an implicit numbered params block or it is actually just a block with no explicit parameters.  I feel this detail alone makes this as a possible point of confusion for new users.  I likely feel it could cause confusion for me if i haven't looked at this node in a year.

Beyond that it add 1-4 bytes to every block when serializing for what is still a somewhat uncommon syntax.  If it does become popular then I feel it also means why put those other fields (parameters+locals) onto a numbered parameters block.

Reply:
It sounds like we agree on a NumberedParameterNode with a single integer field.
I think I'll still desugar it in TruffleRuby but that's no big deal and it makes sense to do it there.

Yeah having null parameters like currently is confusing, because it can mean no parameters at all or numbered parameters. I was thinking the same when touching code before filing https://github.com/ruby/prism/issues/1915. For methods null only means no parameters at all but for blocks it's ambiguous. So a NumberedParameterNode is definitely best to avoid such mistakes.